subsys/suit/stream/stream_filters/src/suit_decompress_filter.c

```diff
@@ -0,0 +1,472 @@
+	*buff_size = ctx.decompressed_image_size > dict_size ?
+				ctx.decompressed_image_size
+				: dict_size;
+
+	return SUIT_PLAT_SUCCESS;
+}
+
+static int close_dictionary(void)
+{
+	return 0;
+}
+
+static size_t write_dictionary(size_t pos, const uint8_t *data, size_t len)
+{
+	suit_plat_err_t res = SUIT_PLAT_SUCCESS;
+
+	if (ctx.curr_image_offset != pos) {
+		ctx.curr_image_offset = pos;
+		res = ctx.out_sink.seek(ctx.out_sink.ctx, ctx.curr_image_offset);
+		if (res != SUIT_PLAT_SUCCESS) {
+			return 0;
+		}
+	}
+
+	res = ctx.out_sink.write(ctx.out_sink.ctx, data, len);
+	ctx.curr_image_offset += len;
+	return res == SUIT_PLAT_SUCCESS ? len : 0;
+}
+
+static size_t read_dictionary(size_t pos, uint8_t *data, size_t len)
+{
+	suit_plat_err_t res = SUIT_PLAT_SUCCESS;
+
+	res = ctx.out_sink_readback(ctx.out_sink.ctx, pos, data, len);
+
+	return res == SUIT_PLAT_SUCCESS ? len : 0;
+}
+
+static suit_plat_err_t used_storage(void *ctx, size_t *size)
+{
+	struct decompress_ctx *decompress_ctx = (struct decompress_ctx *)ctx;
+
+	if ((ctx == NULL) || (size == NULL)) {
+		LOG_ERR("Invalid arguments.");
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	if (decompress_ctx->out_sink.used_storage != NULL) {
+		return decompress_ctx->out_sink.used_storage(decompress_ctx->out_sink.ctx, size);
+	}
+
+	return SUIT_PLAT_ERR_UNSUPPORTED;
+}
+
+static suit_plat_err_t erase(void *ctx)
+{
+	suit_plat_err_t res = SUIT_PLAT_SUCCESS;
+
+	if (ctx != NULL) {
+		struct decompress_ctx *decompress_ctx = (struct decompress_ctx *)ctx;
+		const struct nrf_compress_implementation *codec_impl;
+		size_t size;
+		suit_plat_err_t size_ret;
+
+		codec_impl = decompress_ctx->codec_impl;
+		(void)codec_impl->reset((void *)decompress_ctx->codec_ctx);
+
+		zeroize(decompress_ctx->last_chunk, sizeof(decompress_ctx->last_chunk));
+		decompress_ctx->last_chunk_size = 0;
+
+		size_ret = used_storage(decompress_ctx, &size);
+
+		if (decompress_ctx->out_sink.erase != NULL
+		    && size_ret == SUIT_PLAT_SUCCESS && size != 0) {
+			res = decompress_ctx->out_sink.erase(decompress_ctx->out_sink.ctx);
+		}
+	} else {
+		res = SUIT_PLAT_ERR_INVAL;
+	}
+
+	return res;
+}
+
+static suit_plat_err_t write(void *ctx, const uint8_t *buf, size_t size)
+{
+	suit_plat_err_t res = SUIT_PLAT_SUCCESS;
+	size_t chunk_size;
+	int rc;
+	struct decompress_ctx *decompress_ctx = (struct decompress_ctx *)ctx;
+	uint8_t *output = NULL;
+	size_t output_size = 0;
+	uint32_t processed_size = 0;
+	const struct nrf_compress_implementation *codec_impl;
+
+	if ((ctx == NULL) || (buf == NULL) || (size == 0)) {
+		LOG_ERR("Invalid arguments.");
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	if (!decompress_ctx->in_use) {
+		LOG_ERR("Decrypt filter not initialized.");
+		return SUIT_PLAT_ERR_INVAL;
+	}
+
+	if (decompress_ctx->flushed == true) {
+		LOG_INF("Wrong decompression state.");
+		return SUIT_PLAT_ERR_INCORRECT_STATE;
+	}
+
+	codec_impl = decompress_ctx->codec_impl;
+
+	while (size + decompress_ctx->last_chunk_size > CHUNK_BUFFER_SIZE) {
```

Comment:
You may also use `ARRAY_SIZE(decompress_ctx->last_chunk)` instead of `CHUNK_BUFFER_SIZE` if you find it easier to grasp the idea behind this loop.